devito/dse/graph.py

```diff
@@ -87,148 +101,190 @@ def construct(self, rule):
-    """
-
-    def __init__(self, root, graph, *args, **kwargs):
-        super(Trace, self).__init__(*args, **kwargs)
-        self._root = root
-        self._compute(graph)
-
-    def _compute(self, graph):
-        if self.root not in graph:
-            return
-        to_visit = [(graph[self.root], 0)]
-        while to_visit:
-            temporary, level = to_visit.pop(0)
-            self.__setitem__(temporary.lhs, level)
-            to_visit.extend([(graph[i], level + 1) for i in temporary.reads])
-
-    @property
-    def root(self):
-        return self._root
+        Examples
+        ========
+        Given the following sequence of operations: ::
 
-    @property
-    def length(self):
-        return len(self)
+            t1 = ...
+            t0 = ...
+            u[i, j] = ... v ...
+            u[3, j] = ...
+            v = t0 + t1 + u[z, k]
+            t2 = ...
 
-    def intersect(self, other):
-        return Trace(self.root, {}, [(k, v) for k, v in self.items() if k in other])
+        Assuming ``key == v`` and ``readby is False`` (as by default), return
+        the following list of :class:`Temporary` objects: ::
 
-    def union(self, other):
-        return Trace(self.root, {}, [(k, v) for k, v in self.items() + other.items()])
+            [t1, t0, u[i, j], u[3, j]]
 
+        If ``readby is True``, return: ::
 
-def temporaries_graph(temporaries, scope=0):
+            [v, t2]
+        """
+        if key not in self:
+            return []
+        match = self[key].reads if readby is False else self[key].readby
+        found = []
+        for k, v in self.items():
+            if k in match:
+                found.append(v)
+        return found
+
+    def __getitem__(self, key):
+        if not isinstance(key, slice):
+            return super(TemporariesGraph, self).__getitem__(key)
+        offset = key.step or 0
+        try:
+            start = list(self.keys()).index(key.start) + offset
+        except ValueError:
+            start = 0
+        try:
+            stop = list(self.keys()).index(key.stop) + offset
+        except ValueError:
+            stop = None
+        return TemporariesGraph(islice(list(self.items()), start, stop))
+
+
+def temporaries_graph(temporaries):
     """
-    Create a temporaries graph given a list of :class:`sympy.Eq`.
+    Create a dependency graph given a list of :class:`sympy.Eq`.
     """
 
-    mapper = OrderedDict()
-    Node = namedtuple('Node', ['rhs', 'reads', 'readby'])
-
-    for lhs, rhs in [i.args for i in temporaries]:
-        reads = {i for i in terminals(rhs) if i in mapper}
-        mapper[lhs] = Node(rhs, reads, set())
-        for i in mapper[lhs].reads:
-            assert i in mapper, "Illegal Flow"
-            mapper[i].readby.add(lhs)
+    # Check input is legal and initialize the temporaries graph
+    temporaries = [Temporary(*i.args) for i in temporaries]
+    nodes = [i.lhs for i in temporaries]
+    assert len(set(nodes)) == len(nodes)
```

Comment:
Please move away from using asserts if you ever expect them to catch user errors.